In [ ]:
!pip install -q pyomo


     |████████████████████████████████| 9.6 MB 5.3 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 


Problem 2 \\
$\max \ 0.082x_1+0.036x_2+0.0156x_3+0.033x_4+0.054x_5+0.0125y \\ \text{s.t.} \ 3x_1-3x_2+5x_3-x_4+x_5-y \leq 80 \\ x_3+2x_4+4y \geq 30 \\ \frac{3x_1-1.5x_3+2x_4+1.8x_5+6x_6}{-x_2-x_4+x_5} \leq 3.6 \\
\frac{x_1-2x_2+3x_3+4x_4+5x_5}{-x_1+x_3+x_6-2y}\leq 6 \\ x_1,x_2,x_3,x_4,x_5,x_6 \geq 0 \\ y\in [-0.2,5.5] $

[# case 1 let we are assume that -x2-x4+x5>0
and -x1+x3+x6-x6-2y>0 [link text](https://)

In [ ]:
from pyomo.environ import *


In [ ]:
# create a model1
model2 = ConcreteModel()

In [ ]:
# declare decision variables
model2.x1 = Var(domain=Reals)
model2.x2 = Var(domain=Reals)
model2.x3 = Var(domain=Reals)
model2.x4 = Var(domain=Reals)
model2.x5= Var(domain=Reals)
model2.y = Var(domain=Reals)
model2.x6= Var(domain=Reals)


In [ ]:
# declare objective
model2.cost = Objective(expr = 0.082*model2.x1 + 0.036*model2.x2 + 0.0156*model2.x3 +0.033*model2.x4+0.054*model2.x5+0.0125*model2.y , sense= maximize)

In [ ]:
# declare constraints
model2.constr1 = Constraint(expr = 3*model2.x1-3*model2.x2+5*model2.x3-model2.x4+model2.x5-model2.y <=80)
model2.constr2 = Constraint(expr = model2.x3 +2*model2.x4+4*model2.y>= 30)
model2.constr3 = Constraint(expr = 3*model2.x1 -1.5*model2.x3+2*model2.x4 +1.8*model2.x5+6*model2.x6 +3.6*model2.x2+3.6*model2.x4-3.6*model2.x5 <=0 )
model2.constr4 = Constraint(expr = model2.x1 -2*model2.x2 +3*model2.x3+ 4*model2.x4+5*model2.x5+6*model2.x1-6*model2.x3-6*model2.x6+12*model2.y <=0)
model2.constr6 = Constraint(expr = model2.x1>= 0)
model2.constr7 = Constraint(expr = model2.x2>=0)
model2.constr8 = Constraint(expr = model2.x3>= 0)
model2.constr9= Constraint(expr = model2.x4>= 0)
model2.constr10 = Constraint(expr = model2.x5>= 0)
model2.constr11= Constraint(expr = model2.y>= -0.2)
model2.constr12= Constraint(expr = model2.y<=5.5)
model2.constr13 = Constraint(expr = model2.x6>= 0)


In [ ]:
model2.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fix

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155673 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model2).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 2.12102483443709
  Upper bound: 2.12102483443709
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 31
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.016778230667114258
# ----------------------------------------------------------
#   Solution Information
# ------------

In [ ]:
# display solution
print('\nCost = ', model2.cost())
print('\nDecision Variables')
print('x1 = ', model2.x1())
print('x2 = ', model2.x2())
print('x3 = ', model2.x3())
print('x4 = ', model2.x4())
print('x5 = ', model2.x5())
print('x6 = ', model2.x6())

print('y = ', model2.y())
print('\nConstraints')
print('Constraint 1  = ', model2.constr1())
print('Constraint 2  = ', model2.constr2())
print('Constraint 3  = ', model2.constr3())
print('Constraint 4  = ', model2.constr4())
print('Constraint 6  = ', model2.constr6())
print('Constraint 7  = ', model2.constr7())
print('Constraint 8  = ', model2.constr8())
print('Constraint 9  = ', model2.constr9())
print('Constraint 10  = ', model2.constr10())
print('Constraint 11 = ', model2.constr11())
print('Constraint 12 = ', model2.constr12())
print('Constraint 13 = ', model2.constr13())



Cost =  2.1210248344370837

Decision Variables
x1 =  0.0
x2 =  19.5529801324503
x3 =  24.2384105960265
x4 =  0.0
x5 =  18.9072847682119
x6 =  0.0
y =  1.44039735099338

Constraints
Constraint 1  =  80.00000000000011
Constraint 2  =  30.00000000000002
Constraint 3  =  -9.947598300641403e-14
Constraint 4  =  -4.618527782440651e-14
Constraint 6  =  0.0
Constraint 7  =  19.5529801324503
Constraint 8  =  24.2384105960265
Constraint 9  =  0.0
Constraint 10  =  18.9072847682119
Constraint 11 =  1.44039735099338
Constraint 12 =  1.44039735099338
Constraint 13 =  0.0


 ans 1,$ \max \ \ 0.08x_1+0.036x_2+0.0156x_3+0.033x_4+0.054x_5+0.0125y \\ \textbf{s.t.} \ \ 3x_1-3x_2+5x_3-x_4+x_5-y\leq 180 \\ \ \ \ x_3+2x_4+4y\geq 30 \\ \ \ 3x_1+3.6x_2-1.5x_3+5.6x-4-1.8x_5+6x_6<=0 \\ \ 7x_1-2x_2-3x_3+4x_4+5x_5-6x_6+12y<=0 \\ \ x_1,x_2,x_3,x_4,x_5,x_6,\geq 0 \\ \ y\in \ [-0.2,5.5] $
 # case 1 let we are assume that -x2-x4+x5>0
# and -x1+x3+x6-x6-2y>0 
the solution are exits but 
they are not satisfy of the eqn because  we are assumption is wrong 
Constraint 1  =  80.00000000000011 are active 
Constraint 2  =  30.00000000000002 are active 
Constraint 6  =  0.0 are active 
Constraint 9  =  0.0 are active
Constraint 13 =  0.0 are active 


# case 2 # case 1 let we are assume that -x2-x4+x5<0
and -x1+x3+x6-x6-2y>0 

In [ ]:
!pip install -q pyomo


In [ ]:
from pyomo.environ import *


In [ ]:
# create a model1
model21 = ConcreteModel()

In [ ]:
# declare decision variables
model21.x1 = Var(domain=Reals)
model21.x2 = Var(domain=Reals)
model21.x3 = Var(domain=Reals)
model21.x4 = Var(domain=Reals)
model21.x5= Var(domain=Reals)
model21.y = Var(domain=Reals)
model21.x6= Var(domain=Reals)


In [ ]:
# declare objective
model21.cost = Objective(expr = 0.082*model21.x1 + 0.036*model21.x2 + 0.0156*model21.x3 +0.033*model21.x4+0.054*model21.x5+0.125*model21.y , sense= maximize)

In [ ]:
# declare constraints
model21.constr1 = Constraint(expr = 3*model21.x1-3*model21.x2+5*model21.x3-model21.x4+model21.x5-model21.y <=80)
model21.constr2 = Constraint(expr = model21.x3 +2*model21.x4+4*model21.y>= 30)
model21.constr3 = Constraint(expr = 3*model21.x1 -1.5*model21.x3+2*model21.x4 +1.8*model21.x5+6*model21.x6 +3.6*model21.x2+3.6*model21.x4-3.6*model21.x5 >=0 )
model21.constr4 = Constraint(expr = model2.x1 -2*model21.x2 +3*model21.x3+ 4*model21.x4+5*model21.x5+6*model21.x1-6*model21.x3-6*model21.x6+12*model21.y <=0)
model21.constr6 = Constraint(expr = model21.x1>= 0)
model21.constr7 = Constraint(expr = model21.x2>=0)
model21.constr8 = Constraint(expr = model21.x3>= 0)
model21.constr9= Constraint(expr = model21.x4>= 0)
model21.constr10 = Constraint(expr = model21.x5>= 0)
model21.constr11= Constraint(expr = model21.y>= -0.2)
model21.constr12= Constraint(expr = model21.y<=5.5)
model21.constr13 = Constraint(expr = model21.x6>= 0)


In [ ]:
model21.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fix

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model21).write()

KeyError: ignored

In [ ]:
# display solution
print('\nCost = ', model21.cost())
print('\nDecision Variables')
print('x1 = ', model21.x1())
print('x2 = ', model21.x2())
print('x3 = ', model21.x3())
print('x4 = ', model21.x4())
print('x5 = ', model21.x5())
print('x6 = ', model21.x6())

print('y = ', model21.y())
print('\nConstraints')
print('Constraint 1  = ', model21.constr1())
print('Constraint 2  = ', model21.constr2())
print('Constraint 3  = ', model21.constr3())
print('Constraint 4  = ', model21.constr4())
print('Constraint 6  = ', model21.constr6())
print('Constraint 7  = ', model21.constr7())
print('Constraint 8  = ', model21.constr8())
print('Constraint 9  = ', model21.constr9())
print('Constraint 10  = ', model21.constr10())
print('Constraint 11 = ', model21.constr11())
print('Constraint 12 = ', model21.constr12())
print('Constraint 13 = ', model21.constr12())


# there are feasible solution

In [ ]:
!pip install -q pyomo


# case 3 let we are assume that -x2-x4+x5>0
# and -x1+x3+x6-x6-2y<0

In [ ]:
from pyomo.environ import *


In [ ]:
# create a model1
model22 = ConcreteModel()

In [ ]:
# declare decision variables
model22.x1 = Var(domain=Reals)
model22.x2 = Var(domain=Reals)
model22.x3 = Var(domain=Reals)
model22.x4 = Var(domain=Reals)
model22.x5= Var(domain=Reals)
model22.y = Var(domain=Reals)
model22.x6= Var(domain=Reals)


In [ ]:
# declare objective
model22.cost = Objective(expr = 0.082*model22.x1 + 0.036*model22.x2 + 0.0156*model22.x3 +0.033*model22.x4+0.054*model22.x5+0.125*model22.y , sense= maximize)

In [ ]:
# declare constraints
model22.constr1 = Constraint(expr = 3*model22.x1-3*model22.x2+5*model22.x3-model22.x4+model22.x5-model22.y <=80)
model22.constr2 = Constraint(expr = model22.x3 +2*model22.x4+4*model22.y>= 30)
model22.constr3 = Constraint(expr = 3*model22.x1 -1.5*model2.x3+2*model2.x4 +1.8*model2.x5+6*model2.x6 +3.6*model2.x2+3.6*model2.x4-3.6*model2.x5 <=0 )
model22.constr4 = Constraint(expr = model2.x1 -2*model2.x2 +3*model2.x3+ 4*model2.x4+5*model2.x5+6*model2.x1-6*model2.x3-6*model2.x6+12*model2.y >=0)
model22.constr6 = Constraint(expr = model22.x1>= 0)
model2.constr7 = Constraint(expr = model22.x2>=0)
model22.constr8 = Constraint(expr = model2.x3>= 0)
model22.constr9= Constraint(expr = model22.x4>= 0)
model22.constr10 = Constraint(expr = model22.x5>= 0)
model22.constr11= Constraint(expr = model22.y>= -0.2)
model22.constr12= Constraint(expr = model22.y<=5.5)
model22.constr13 = Constraint(expr = model22.x6>= 0)


    'pyomo.core.base.constraint.ScalarConstraint'>) on block unknown with a
    new Component (type=<class
    'pyomo.core.base.constraint.AbstractScalarConstraint'>). This is usually
    indicative of a modelling error. To avoid this warning, use
    block.del_component() and block.add_component().


In [ ]:
model22.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fix

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model22).write()

KeyError: ignored

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model22).write()

KeyError: ignored

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model22).write()

KeyError: ignored

In [ ]:
# display solution
print('\nCost = ', model22.cost())
print('\nDecision Variables')
print('x1 = ', model22.x1())
print('x2 = ', model22.x2())
print('x3 = ', model22.x3())
print('x4 = ', model22.x4())
print('x5 = ', model22.x5())
print('x6 = ', model22.x6())

print('y = ', model22.y())
print('\nConstraints')
print('Constraint 1  = ', model22.constr1())
print('Constraint 2  = ', model22.constr2())
print('Constraint 3  = ', model22.constr3())
print('Constraint 4  = ', model22.constr4())
print('Constraint 6  = ', model22.constr6())
print('Constraint 7  = ', model22.constr7())
print('Constraint 8  = ', model22.constr8())
print('Constraint 9  = ', model22.constr9())
print('Constraint 10  = ', model22.constr10())
print('Constraint 11 = ', model22.constr11())
print('Constraint 12 = ', model22.constr12())
print('Constraint 13 = ', model22.constr12())


# there are feasible solution

# case 4 let we are assume that -x2-x4+x5<0
and -x1+x3+x6-x6-2y<0

In [ ]:
!pip install -q pyomo


In [ ]:
from pyomo.environ import *


In [ ]:
# create a model1
model23 = ConcreteModel()

In [ ]:
# declare decision variables
model23.x1 = Var(domain=Reals)
model23.x2 = Var(domain=Reals)
model23.x3 = Var(domain=Reals)
model23.x4 = Var(domain=Reals)
model23.x5= Var(domain=Reals)
model23.y = Var(domain=Reals)
model23.x6= Var(domain=Reals)


In [ ]:
# declare objective
model23.cost = Objective(expr = 0.082*model23.x1 + 0.036*model23.x2 + 0.0156*model23.x3 +0.033*model23.x4+0.054*model23.x5+0.125*model23.y , sense= maximize)

In [ ]:
# declare constraints
model23.constr1 = Constraint(expr = 3*model23.x1-3*model23.x2+5*model23.x3-model23.x4+model23.x5-model23.y <=80)
model23.constr2 = Constraint(expr = model23.x3 +2*model23.x4+4*model23.y>= 30)
model23.constr3 = Constraint(expr = 3*model23.x1 -1.5*model23.x3+2*model23.x4 +1.8*model23.x5+6*model23.x6 +3.6*model23.x2+3.6*model23.x4-3.6*model23.x5 >=0 )
model23.constr4 = Constraint(expr = model23.x1 -2*model23.x2 +3*model23.x3+ 4*model23.x4+5*model23.x5+6*model23.x1-6*model23.x3-6*model23.x6+12*model23.y >=0)
model23.constr6 = Constraint(expr = model23.x1>= 0)
model23.constr7 = Constraint(expr = model23.x2>=0)
model23.constr8 = Constraint(expr = model23.x3>= 0)
model23.constr9= Constraint(expr = model23.x4>= 0)
model23.constr10 = Constraint(expr = model23.x5>= 0)
model23.constr11= Constraint(expr = model23.y>= -0.2)
model23.constr12= Constraint(expr = model23.y<=5.5)
model23.constr13 = Constraint(expr = model23.x6>= 0)



In [ ]:
model23.pprint()

7 Var Declarations
    x1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x5 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    x6 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    y : Size=1, Index=None
        Key  : Lower : Value : Upper : Fix

In [ ]:
!apt-get install -y -qq glpk-utils

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model23).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 13
  Number of variables: 8
  Number of nonzeros: 31
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.0170137882232666


In [ ]:
# display solution
print('\nCost = ', model21.cost())
print('\nDecision Variables')
print('x1 = ', model23.x1())
print('x2 = ', model23.x2())
print('x3 = ', model23.x3())
print('x4 = ', model23.x4())
print('x5 = ', model23.x5())
print('x6 = ', model23.x6())
print('y = ', model23.y())
print('\nConstraints')
print('Constraint 1  = ', model23.constr1())
print('Constraint 2  = ', model23.constr2())
print('Constraint 3  = ', model23.constr3())
print('Constraint 4  = ', model23.constr4())
print('Constraint 6  = ', model23.constr6())
print('Constraint 7  = ', model23.constr7())
print('Constraint 8  = ', model23.constr8())
print('Constraint 9  = ', model23.constr9())
print('Constraint 10  = ', model23.constr10())
print('Constraint 11 = ', model23.constr11())
print('Constraint 12 = ', model23.constr12())
print('Constraint 13 = ', model23.constr12())


ERROR: evaluating object as numeric value: x1
        (object: <class 'pyomo.core.base.var.ScalarVar'>)
    No value for uninitialized NumericValue object x1


ValueError: ignored

there is no feasible solution